File: go/14_text_to_speech.go
package main

/*
Example 14: Text-to-Speech with OpenAI

Demonstrates:
- Converting text to spoken audio
- 11 different voices
- Instructions and speed parameters
- Multi-language support

Note: Uses Go standard library only
*/

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
)

func textToSpeech(text, voice string, options map[string]interface{}) ([]byte, error) {
	requestData := map[string]interface{}{
		"model": "gpt-4o-mini-tts",
		"input": text,
		"voice": voice,
	}
	
	for k, v := range options {
		requestData[k] = v
	}
	
	requestBody, _ := json.Marshal(requestData)
	
	req, err := http.NewRequest("POST", "https://api.openai.com/v1/audio/speech", bytes.NewBuffer(requestBody))
	if err != nil {
		return nil, err
	}
	
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", os.Getenv("OPENAI_API_KEY")))
	
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	
	return io.ReadAll(resp.Body)
}

func main() {
	if os.Getenv("OPENAI_API_KEY") == "" {
		fmt.Fprintln(os.Stderr, "Error: OPENAI_API_KEY not set")
		fmt.Fprintln(os.Stderr, "Get your key from: https://platform.openai.com")
		os.Exit(1)
	}
	
	fmt.Println("========================================")
	fmt.Println("Text-to-Speech with OpenAI")
	fmt.Println("========================================\n")
	
	// Demo 1: Basic TTS
	text := "Hello! This is an example of text-to-speech synthesis using OpenAI's affordable TTS model."
	audio, err := textToSpeech(text, "alloy", nil)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
	
	os.WriteFile("demo1_alloy.mp3", audio, 0644)
	fmt.Printf("✓ Audio created: demo1_alloy.mp3 (%d bytes)\n\n", len(audio))
	
	// Demo 2: All voices
	fmt.Println("Generating all 11 voices...\n")
	voices := []string{"alloy", "ash", "ballad", "coral", "echo", "fable", 
	                   "onyx", "nova", "sage", "shimmer", "verse"}
	
	comparisonText := "Welcome to OpenAI's text-to-speech demonstration."
	for _, voice := range voices {
		audio, _ := textToSpeech(comparisonText, voice, nil)
		filename := fmt.Sprintf("voice_%s.mp3", voice)
		os.WriteFile(filename, audio, 0644)
		fmt.Printf("✓ %s: %s\n", voice, filename)
	}
	
	fmt.Println("\n========================================")
	fmt.Println("Cost Analysis")
	fmt.Println("========================================\n")
	fmt.Println("Pricing: $0.60/1M input tokens, $12/1M output tokens")
	fmt.Printf("This demo: ~%d chars = $%.8f input\n\n", len(text), float64(len(text))*0.60/1000000)
	fmt.Println("Much more affordable than alternatives!\n")
}

File: python/14_text_to_speech.py
#!/usr/bin/env python3
"""
Example 14: Text-to-Speech with OpenAI

Demonstrates:
- Converting text to spoken audio with OpenAI's TTS
- 11 different voices and instructions parameter
- Speed control and multi-language support

Note: Uses Python standard library only
"""

import http.client
import json
import os
import sys

if not os.environ.get('OPENAI_API_KEY'):
    print('Error: OPENAI_API_KEY not set', file=sys.stderr)
    print('Get your key from: https://platform.openai.com', file=sys.stderr)
    sys.exit(1)

def text_to_speech(text, voice, **kwargs):
    """Call OpenAI TTS API"""
    conn = http.client.HTTPSConnection('api.openai.com')
    
    request_data = {
        'model': 'gpt-4o-mini-tts',
        'input': text,
        'voice': voice,
        **kwargs
    }
    
    body = json.dumps(request_data)
    headers = {
        'Content-Type': 'application/json',
        'Authorization': f'Bearer {os.environ["OPENAI_API_KEY"]}'
    }
    
    conn.request('POST', '/v1/audio/speech', body, headers)
    response = conn.getresponse()
    audio_data = response.read()
    conn.close()
    
    return audio_data

def main():
    print('========================================')
    print('Text-to-Speech with OpenAI')
    print('========================================\n')
    
    # Demo 1: Basic TTS
    text = "Hello! This is an example of text-to-speech synthesis using OpenAI's affordable TTS model."
    audio = text_to_speech(text, 'alloy')
    
    with open('demo1_alloy.mp3', 'wb') as f:
        f.write(audio)
    
    print(f'✓ Audio created: demo1_alloy.mp3 ({len(audio)} bytes)\n')
    
    # Demo 2: All voices
    print('Generating all 11 voices...\n')
    voices = ['alloy', 'ash', 'ballad', 'coral', 'echo', 'fable', 
              'onyx', 'nova', 'sage', 'shimmer', 'verse']
    
    comparison_text = "Welcome to OpenAI's text-to-speech demonstration."
    for voice in voices:
        audio = text_to_speech(comparison_text, voice)
        filename = f'voice_{voice}.mp3'
        with open(filename, 'wb') as f:
            f.write(audio)
        print(f'✓ {voice}: {filename}')
    
    print('\n' + '='*40)
    print('Cost Analysis')
    print('='*40 + '\n')
    print('Pricing: $0.60/1M input tokens, $12/1M output tokens')
    print(f'This demo: ~{len(text)} chars = ${len(text)*0.60/1000000:.8f} input\n')
    print('Much more affordable than alternatives!\n')

if __name__ == '__main__':
    try:
        main()
    except Exception as error:
        print(f'Error: {error}', file=sys.stderr)
        sys.exit(1)
